ros_buildfarm/templates/release/rpm/binarypkg_job.xml.em

```diff
@@ -147,42 +147,14 @@ but disabled since the package is blacklisted (or not whitelisted) in the config
     'builder_shell',
     script='\n'.join([
         'echo "# BEGIN SECTION: Upload binaryrpm"',
-        'export PYTHONPATH=$WORKSPACE/ros_buildfarm:$PYTHONPATH',
-        "ls binarypkg/*.rpm | grep -v -e 'src\.rpm$' -e '-debug\(info\|source\)-' > binarypkg/upload_list.txt && " +
-        'xargs -a binarypkg/upload_list.txt' +
-        ' python3 -u $WORKSPACE/ros_buildfarm/scripts/release/rpm/upload_package.py' +
-        ' --pulp-resource-record binarypkg/upload_record.txt',
-        'echo "# END SECTION"',
-        'echo "# BEGIN SECTION: Upload debug symbols"',
-        'export PYTHONPATH=$WORKSPACE/ros_buildfarm:$PYTHONPATH',
-        "ls binarypkg/*.rpm | grep -e '-debug\(info\|source\)-' > binarypkg/upload_list_debug.txt && " +
-        'xargs -a binarypkg/upload_list_debug.txt' +
-        ' python3 -u $WORKSPACE/ros_buildfarm/scripts/release/rpm/upload_package.py' +
-        ' --pulp-resource-record binarypkg/upload_record_debug.txt',
+        'find binarypkg -mindepth 1 -maxdepth 1 -type f -name "*.rpm" -not -name "*.src.rpm" -fprint binarypkg/rpm_upload_args.txt -fprintf binarypkg/rpm_import_args.txt "--import=/tmp/upload-${BUILD_TAG}/%f\\n"',
+        'ssh $UPLOAD_HOST -- mkdir -p /tmp/upload-${BUILD_TAG}/',
+        'xargs -a binarypkg/rpm_upload_args.txt -I % scp % $UPLOAD_HOST:/tmp/upload-${BUILD_TAG}/',
+        'xargs -a binarypkg/rpm_import_args.txt ssh $UPLOAD_HOST -- createrepo-agent /var/repos/%s/building/%s/ --arch %s --invalidate-family --invalidate-dependants' % (os_name, os_code_name, arch),
+        'ssh $UPLOAD_HOST -- rm -fr /tmp/upload-${BUILD_TAG}/',
```

Comment:
There are four separate SSH connections being opened here. The overhead of the excess connections accounts for the majority of time it takes to do most package uploads entirely. It would be nice to either enable OpenSSH features that allow us to re-use connections here or switch to an approach that can both upload and execute a script on the host in the same connection. I've investigated the latter and found ways to make it happen, but none that I found straightforward or well supported.

Reply:
It seems like configuring a persistent control socket with `ControlMaster` and `ControlPersist`  could get us what we need, if we do open a persistent control socket we just need to take care that we close it reliably on each job's completion.

https://manpages.ubuntu.com/manpages/focal/en/man5/ssh_config.5.html